### PR TITLE
fix(npm): Prevent unnecessary hash pinning in lock file constraint

### PR DESCRIPTION
**Pull request by xlgmokha**
# Why is this needed?

To make sure that Dependabot does not pin to a specific git hash when the `package.json` has not pinned to one.

## What does this do?

It updates the npm helper to check for a pinned hash in git based version constraints.

Fixes https://github.com/dependabot/dependabot-core/issues/3813

### Screenshots

Before:

```bash
[dependabot-core-dev] ~/dependabot-core $ ./bin/dry-run.rb npm_and_yarn xlgmokha/hypixel-translators-bot --dep discord.js --branch core-3813
Resolving dependencies...
warning: parser/current is loading parser/ruby26, which recognizes
warning: 2.6.7-compliant syntax, but you are running 2.6.6.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
=> fetching dependency files
=> dumping fetched dependency files: ./dry-run/xlgmokha/hypixel-translators-bot/core-3813/
=> parsing dependency files
=> updating 1 dependencies: discord.js

=== discord.js (ab82cafcde0ee259a32ef14303c1b4a64dea8fae)
 => checking for updates 1/1
 => latest available version is 7b85a7259f563ab14ae6c0a665a3dd43c486fde4
 => latest allowed version is 7b85a7259f563ab14ae6c0a665a3dd43c486fde4
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating discord.js from ab82cafcde0ee259a32ef14303c1b4a64dea8fae to 7b85a7259f563ab14ae6c0a665a3dd43c486fde4

    ± package-lock.json
    ~~~
    37a38,42
    >     "@sapphire/async-queue": {
    >       "version": "1.1.2",
    >       "resolved": "https://registry.npmjs.org/@sapphire/async-queue/-/async-queue-1.1.2.tgz",
    >       "integrity": "sha512-NkR7AzC9uyb++tMIZgG4X0ci8JM1rnvNmKbLwY42RgotRV8JGUntZ7ZpR7MN7p5zPlVdKo/YmOqcCCsBJ6LNuw=="
    >     },
    88d92
    <       "dev": true,
    308,309c312
    <       "integrity": "sha512-hNC38R9ZF4uaujaZQtQfm5CdQO58uhdkoHQAVvMfIL0LgOSZeW575W8H6upngQOuoxWd8tiRII3LLJm9zuQKYg==",
    <       "dev": true
    ---
    >       "integrity": "sha512-hNC38R9ZF4uaujaZQtQfm5CdQO58uhdkoHQAVvMfIL0LgOSZeW575W8H6upngQOuoxWd8tiRII3LLJm9zuQKYg=="
    312,313c315,316
    <       "version": "github:discordjs/discord.js#ab82cafcde0ee259a32ef14303c1b4a64dea8fae",
    <       "from": "github:discordjs/discord.js",
    ---
    >       "version": "github:discordjs/discord.js#7b85a7259f563ab14ae6c0a665a3dd43c486fde4",
    >       "from": "github:discordjs/discord.js#7b85a7259f563ab14ae6c0a665a3dd43c486fde4",
    316a320,321
    >         "@sapphire/async-queue": "^1.1.2",
    >         "@types/ws": "^7.4.4",
    317a323
    >         "discord-api-types": "^0.18.1",
    319c325
    <         "prism-media": "^1.2.2",
    ---
    >         "prism-media": "^1.2.9",
    321c327,334
    <         "ws": "^7.3.1"
    ---
    >         "ws": "^7.4.6"
    >       },
    >       "dependencies": {
    >         "ws": {
    >           "version": "7.4.6",
    >           "resolved": "https://registry.npmjs.org/ws/-/ws-7.4.6.tgz",
    >           "integrity": "sha512-YmhHDO4MzaDLB+M9ym/mDA5z0naX8j7SIlT8f8z+I0VtzsRbekxEutHSme7NPS2qE8StCYQNUnfWdXta/Yu85A=="
    >         }
    ~~~
```

After:

```bash
$ ./bin/dry-run.rb npm_and_yarn xlgmokha/hypixel-translators-bot --dep 'discord.js' --branch core-3813
warning: parser/current is loading parser/ruby26, which recognizes
warning: 2.6.7-compliant syntax, but you are running 2.6.6.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
=> fetching dependency files
=> dumping fetched dependency files: ./dry-run/xlgmokha/hypixel-translators-bot/core-3813/
=> parsing dependency files
=> updating 1 dependencies: discord.js

=== discord.js (ab82cafcde0ee259a32ef14303c1b4a64dea8fae)
 => checking for updates 1/1
 => latest available version is 51551f544b80d7d27ab0b315da01dfc560b2c115
 => latest allowed version is 51551f544b80d7d27ab0b315da01dfc560b2c115
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating discord.js from ab82cafcde0ee259a32ef14303c1b4a64dea8fae to 51551f544b80d7d27ab0b315da01dfc560b2c115

    ± package-lock.json
    ~~~
    37a38,42
    >     "@sapphire/async-queue": {
    >       "version": "1.1.2",
    >       "resolved": "https://registry.npmjs.org/@sapphire/async-queue/-/async-queue-1.1.2.tgz",
    >       "integrity": "sha512-NkR7AzC9uyb++tMIZgG4X0ci8JM1rnvNmKbLwY42RgotRV8JGUntZ7ZpR7MN7p5zPlVdKo/YmOqcCCsBJ6LNuw=="
    >     },
    88d92
    <       "dev": true,
    308,309c312
    <       "integrity": "sha512-hNC38R9ZF4uaujaZQtQfm5CdQO58uhdkoHQAVvMfIL0LgOSZeW575W8H6upngQOuoxWd8tiRII3LLJm9zuQKYg==",
    <       "dev": true
    ---
    >       "integrity": "sha512-hNC38R9ZF4uaujaZQtQfm5CdQO58uhdkoHQAVvMfIL0LgOSZeW575W8H6upngQOuoxWd8tiRII3LLJm9zuQKYg=="
    312c315
    <       "version": "github:discordjs/discord.js#ab82cafcde0ee259a32ef14303c1b4a64dea8fae",
    ---
    >       "version": "github:discordjs/discord.js#fdad14099779e61cb84dcd1cb2497e0e853a6144",
    316a320,321
    >         "@sapphire/async-queue": "^1.1.2",
    >         "@types/ws": "^7.4.4",
    317a323
    >         "discord-api-types": "^0.18.1",
    319c325
    <         "prism-media": "^1.2.2",
    ---
    >         "prism-media": "^1.2.9",
    321c327,334
    <         "ws": "^7.3.1"
    ---
    >         "ws": "^7.4.6"
    >       },
    >       "dependencies": {
    >         "ws": {
    >           "version": "7.4.6",
    >           "resolved": "https://registry.npmjs.org/ws/-/ws-7.4.6.tgz",
    >           "integrity": "sha512-YmhHDO4MzaDLB+M9ym/mDA5z0naX8j7SIlT8f8z+I0VtzsRbekxEutHSme7NPS2qE8StCYQNUnfWdXta/Yu85A=="
    >         }
    ~~~
```

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by ImRodry** [comment]
Thanks for this! I see that you committed our package and package-lock files, is this intentional?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb [hunk 1]
```diff
@@ -256,7 +257,7 @@ def requirement_for(requirement)
         return requirement unless git_url?(requirement)
 
         details = requirement.match(GIT_URL_REGEX).named_captures
-        details["semver"]
+        details["semver"] || (requirement.start_with?("github:") ? requirement : nil)
```
Comment (feelepxyz) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:260]: Do we need to change this? Wondering if we should leave this as is, being `nil` if it doesn't match semver. Doesn't look like we actually rely on this change anywhere and might cause issues upstream issues when creating pull requests.

File: common/lib/dependabot/file_parsers/base/dependency_set.rb [hunk 2]
```diff
@@ -21,6 +21,7 @@ def initialize(dependencies = [], case_sensitive: false)
         attr_reader :dependencies
 
         def <<(dep)
+          return if dep.nil?
```
Comment (feelepxyz) [inline common/lib/dependabot/file_parsers/base/dependency_set.rb:24]: Is this change required? Less sure about changing the common classes unless we need to and make use of the change across each package manager to not introduce subtle differences in implementation.

File: npm_and_yarn/helpers/lib/npm6/updater.js [hunk 3]
```diff
@@ -138,10 +138,10 @@ function installArgs(
...
       return `${depName}@${existingVersionRequirement}`;
     } else {
-      return `${depName}@${existingVersionRequirement.replace(
-        /#.*/,
-        ""
-      )}#${desiredVersion}`;
+      if (existingVersionRequirement.includes("#")) {
+        return `${depName}@${existingVersionRequirement.replace(/#.*/, "")}#${desiredVersion}`;
+      }
+      return `${depName}@${existingVersionRequirement}`;
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/lib/npm6/updater.js:144]: Nice detective work on finding this fix! 🙌

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb [hunk 4]
```diff
@@ -55,21 +55,22 @@ def manifest_dependencies
...
-              requirement = "*" if requirement == ""
-              dep = build_dependency(
-                file: file, type: type, name: name, requirement: requirement
+              dependency_set << build_dependency(
+                file: file,
+                type: type,
+                name: name,
+                requirement: requirement.empty? ? "*" : requirement
               )
-              dependency_set << dep if dep
```
Comment (feelepxyz) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:72]: Related to the change to the dependency_set, thoughts on keeping this behaviour here so we don't need to change the common class?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb [hunk 5]
```diff
@@ -151,11 +151,12 @@ def recursively_fetch_npm_lock_dependencies(object_with_dependencies)
 
           object_with_dependencies.
             fetch("dependencies", {}).each do |name, details|
-              next unless semver_version_for(details["version"])
+              version = semver_version_for(details["version"])
```
Comment (feelepxyz) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:154]: 👍

File: npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "hypixel-translators",
```
Comment (feelepxyz) [inline npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json:2]: This would be good to strip down to basic example, maybe just including `discord.js` dependency. An easy way of generating an example with npm is running `yes "" | npm init && npm install github:discordjs/discord.js#ab82cafcde0ee259a32ef14303c1b4a64dea8fae --save --lockfile-only`, if you then remove the `#sha` from `package.json` and re-run `npm install` it should then update the lockfile `from` to match keeping the old sha `version`

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (feelepxyz) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:260]: I think this steer is a good idea but unsure of just changing it here for requirements that specifically match `github:`. Might be a good idea to figure out if we could always insert the requirement here but feels out of scope for this change.

**Comment by xlgmokha** [comment]
> Thanks for this! I see that you committed our package and package-lock files, is this intentional?

No, this is temporary until I could wrap my head around the problem. I'll tidy up the fixture file before submitting for review.

File: common/lib/dependabot/file_parsers/base/dependency_set.rb (hunk 2 above)
Comment (xlgmokha) [reply common/lib/dependabot/file_parsers/base/dependency_set.rb:24]: > Is this change required?

Nope. I'll revert it.

File: npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json (hunk 6 above)
Comment (ImRodry) [reply npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json:2]: You shouldn't need to install with the #sha to begin with though?

Comment (feelepxyz) [reply npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json:2]: > You shouldn't need to install with the #sha to begin with though?

I think we want to make sure the installed sha is old and can be updated

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (xlgmokha) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: > Do we need to change this?

Yes. I believe this this is the root cause of the issue. A `nil` constraint doesn't provide `npm` with enough information. More info here https://github.com/dependabot/dependabot-core/issues/3813#issuecomment-853458082.

I would prefer to change this line to:

```ruby
  details["semver"] || requirement
```

This would ensure that we always pass the original constraint down to npm to complete an install using the constraint provided in the manifest. However, I'm unsure of the other consequences of this change. Because of that I chose to be defensive and try to restrict this to only [GitHub repository style URLs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository) for the time being. Without this change the defect is consistently reproduced.

File: npm_and_yarn/helpers/lib/npm6/updater.js (hunk 3 above)
Comment (xlgmokha) [reply npm_and_yarn/helpers/lib/npm6/updater.js:144]: Thanks. This was a fun side quest.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 5 above)
Comment (xlgmokha) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:154]: I noticed a few places where we could memoize work in the npm code. Since npm seems to be one of our slower ecosystems, it felt like a safe and beneficial change to make. Although, I recognize that this change isn't necessary for resolving this particular defect.

File: npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json (hunk 6 above)
Comment (xlgmokha) [reply npm_and_yarn/spec/fixtures/projects/npm6/ghpr_no_hash_pinning/package.json]: I updated the fixture to make sure that the declared constraint isn't pinned to a sha and that the `from` field in the lockfile matches this constraint. Please let me know if I've missed anything else.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (feelepxyz) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: Ah not sure this gets used in the npm update, the requirement is extracted from the `package.json` file here https://github.com/dependabot/dependabot-core/blob/0b514b24ff8a201e379157a22a50107f7d86af27/npm_and_yarn/helpers/lib/npm6/updater.js#L52

Confusingly we also do this here https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L252-L257 🤔

I think this might need to broader refactors to clean up, I definitely think it would be a lot better to pass this from the file parser instead of re-extracting it in the lockfile updater and native helper.

Comment (xlgmokha) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: I am looking for more context around how the requirement is used in the downstream processing. Please do tell me what you think and if you have any alternative ideas that I can/should pursue.

Comment (feelepxyz) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: Does the dry-run update work if you revert this change? If so would be up for shipping this, I have a feeling this change doesn't have an impact on the update so might just cause issues in the pr creator where we use the requirement to build the PR title/body

Comment (feelepxyz) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: > Without this change the defect is consistently reproduced

I tested this dry-run without this change and seems to work but might be missing something else?

```
bin/dry-run.rb npm_and_yarn Hypixel-Translators/hypixel-translators-bot --dep=discord.js --cache=files --commit=e2bfc6768492d32f22bdc6a732237315b6b22dd2
```

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater_spec.rb [hunk 7]
```diff
@@ -500,4 +500,42 @@
...
+          type: "git",
+          url: "https://github.com/discordjs/discord.js",
+          branch: nil,
+          ref: "master"
+        }
+      }]
+    end
+
+    specify { expect(subject["dependencies"]["discord.js"]["version"]).to match(/github:discordjs\/discord.js#[0-9a-z]{40}/) }
+    specify { expect(subject["dependencies"]["discord.js"]["from"]).to eq("github:discordjs/discord.js") }
```
Comment (feelepxyz) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater_spec.rb:539]: Nit, looks like this would run the update twice to check each subject. Thoughts on using a single `it` here instead and check both attributes?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (xlgmokha) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:261]: > Ah not sure this gets used in the npm update, the requirement is extracted from the `package.json` file here
> 
> https://github.com/dependabot/dependabot-core/blob/0b514b24ff8a201e379157a22a50107f7d86af27/npm_and_yarn/helpers/lib/npm6/updater.js#L52
> 
> Confusingly we also do this here https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L252-L257 🤔
> 
> I think this might need to broader refactors to clean up, I definitely think it would be a lot better to pass this from the file parser instead of re-extracting it in the lockfile updater and native helper.

You're right. We don't need this. I just re-tested with and without and both appear to leave the `from` field in the desired state. I'll remove this change. Thank you for spotting this @feelepxyz. 🙇

**Review by feelepxyz — APPROVED** [review]
🙌 🚢

**Review by feelepxyz — APPROVED** [review]

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater_spec.rb [hunk 8]
```diff
@@ -316,7 +316,7 @@
         }]
       end
       let(:previous_version) { "d5ac0584ee9ae7bd9288220a39780f155b9ad4c8" }
-      let(:version) { "0c6b15a88bc10cd47f67a09506399dfc9ddc075d" }
+      let(:version) { "98e8ff1da1a89f93d1397a24d7413ed15421c139" }
```
Comment (feelepxyz) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater_spec.rb:319]: Oh why did this change?

Comment (feelepxyz) [reply npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater_spec.rb:319]: Ah this is the latest commit? hm not sure we should change this variable for all tests as we then loose code paths testing updating to specific commits vs latest when there's no reference. I think we still want some tests that try to update to a specific old commit, e.g. pointing to a tag.

Comment (xlgmokha) [reply npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater_spec.rb]: > Oh why did this change?

The fixtures in those tests do not pin to a hash (e.g. `"is-number": "jonschlinkert/is-number"`) so the npm install step fetches the latest hash for the default branch of that repo.

> Ah this is the latest commit?

Yep. 

> hm not sure we should change this variable for all tests as we then loose code paths testing updating to specific commits vs latest when there's no reference

🤔 Can I update the constraint in the package.json for those fixtures to pin to a fixed version?

**Comment by xlgmokha** [comment]
@feelepxyz would you mind double checking the latest commit to this PR?

**Review by feelepxyz — APPROVED** [review]
sgtm 👍 seems ok to punt on copying the git dep over to dependabot-fixtures, could just do it when/if it ever flakes
